internal/output/vertical.go

```diff
@@ -0,0 +1,243 @@
+	vulns := make([]models.Vulnerability, 0)
+
+	for _, group := range pkg.Groups {
+		if group.IsCalled() != called {
+			continue
+		}
+
+		for _, id := range group.IDs {
+			for _, v := range pkg.Vulnerabilities {
+				if v.ID == id {
+					vulns = append(vulns, v)
+				}
+			}
+		}
+	}
+
+	return vulns
+}
+
+func printVerticalVulnerabilitiesForPackages(result models.PackageSource, out io.Writer, printingCalled bool) {
+	for _, pkg := range result.Packages {
+		vulns := collectVulns(pkg, printingCalled)
+
+		if len(vulns) == 0 {
+			continue
+		}
+
+		state := "uncalled"
+		if printingCalled {
+			state = "known"
+		}
+
+		fmt.Fprintf(out,
+			"  %s %s\n",
+			text.FgYellow.Sprintf("%s@%s", pkg.Package.Name, pkg.Package.Version),
+			text.FgRed.Sprintf("has the following %s vulnerabilities:", state),
+		)
+
+		for _, vulnerability := range vulns {
+			fmt.Fprintf(out,
+				"    %s %s\n",
+				text.FgCyan.Sprintf("%s:", vulnerability.ID),
+				describe(vulnerability),
+			)
+		}
+	}
+}
+
+func printVerticalVulnerabilities(result models.PackageSource, out io.Writer) {
+	countCalled, countUncalled := countVulnerabilities(result)
+
+	if countCalled == 0 && countUncalled == 0 {
+		fmt.Fprintf(
+			out,
+			"  %s\n",
+			text.FgGreen.Sprintf("no known vulnerabilities found"),
+		)
+
+		return
+	}
+
+	if countCalled > 0 {
+		fmt.Fprintln(out)
+
+		printVerticalVulnerabilitiesForPackages(result, out, true)
+		printVerticalVulnerabilitiesCountSummary(countCalled, "known", result.Source.Path, out)
+	}
+
+	if countUncalled > 0 {
+		fmt.Fprintln(out)
+
+		printVerticalVulnerabilitiesForPackages(result, out, false)
+		printVerticalVulnerabilitiesCountSummary(countUncalled, "uncalled", result.Source.Path, out)
+	}
+}
+
+func printVerticalLicenseViolations(result models.PackageSource, out io.Writer) {
+	count := countLicenseViolations(result)
+
+	if count == 0 {
+		fmt.Fprintf(
+			out,
+			"  %s\n",
+			text.FgGreen.Sprintf("no license violations found"),
+		)
+
+		return
+	}
+
+	fmt.Fprintf(out, "\n  %s\n", text.FgRed.Sprintf("license violations found:"))
+
+	for _, pkg := range result.Packages {
+		if len(pkg.LicenseViolations) == 0 {
+			continue
+		}
+
+		violations := make([]string, len(pkg.LicenseViolations))
+		for i, l := range pkg.LicenseViolations {
+			violations[i] = string(l)
+		}
+
+		fmt.Fprintf(out,
+			"    %s (%s)\n",
+			text.FgYellow.Sprintf("%s@%s", pkg.Package.Name, pkg.Package.Version),
+			text.FgCyan.Sprintf(strings.Join(violations, ", ")),
+		)
+	}
+
+	fmt.Fprintf(out, "\n  %s\n",
+		text.FgRed.Sprintf(
+			"%d license %s found in %s",
+			count,
+			Form(count, "violation", "violations"),
+			result.Source.Path,
+		),
+	)
+}
+
+func countVulnerabilities(result models.PackageSource) (called int, uncalled int) {
+	for _, pkg := range result.Packages {
+		for _, group := range pkg.Groups {
+			for _, id := range group.IDs {
+				for _, v := range pkg.Vulnerabilities {
+					if v.ID == id {
+						if group.IsCalled() {
```

Comment:
it's also kind of indented. Should we use some extra data structures to track vulnerability count?

Reply:
I'd love to have this less intended but I don't really think there's a way to do that which wouldn't just be smearing the indenting around (/ replacing it with different complexity).

Not saying we should do this everywhere, but personally I view it as a natural downside to the currently complex-but-necessary shape of the data structure, and something to review as/when there's more need for this kind of stuff

i.e. v2 might involve some reshaping to this structure that lets us improve this, or a new property might come along that makes it useful to have some functions that could also be used here